### PR TITLE
Sign in

### PR DESCRIPTION
**Pull request by phong016688**

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -22,4 +23,4 @@
             </intent-filter>
         </activity>
     </application>
-</manifest>
+</manifest>
```
Comment (onstonboy) [inline app/src/main/AndroidManifest.xml]: add blank line

File: app/src/main/java/com/ccc/chatapp/repositories/UserRepository.kt [hunk 2]
```diff
@@ -24,7 +27,12 @@ class UserRepositoryImpl(sharedPrefsApi: SharedPrefsApi) : UserRepository {
         return Single.create { emitter ->
             mAuth.signInWithEmailAndPassword(username, password)
                 .addOnCompleteListener { task ->
-                    if (task.isSuccessful) {
+                    if (task.isSuccessful && isUserLogged()) {
+                        val dbRef = mFirestore.collection("user")
+                            .document(mAuth.currentUser!!.uid)
```
Comment (onstonboy) [inline app/src/main/java/com/ccc/chatapp/repositories/UserRepository.kt]: mAuth.currentUser!! -> Hạn chế dùng !! (force-wrap). nên check !null trước khi dùng.
suggest:
val currentUser = mAuth.currentUser ?: return

File: app/src/main/java/com/ccc/chatapp/screens/chat/ChatActivity.kt [hunk 3]
```diff
@@ -48,7 +48,7 @@ class ChatActivity : AppCompatActivity(), ChatView {
     }
 
     override fun onOptionsItemSelected(item: MenuItem?): Boolean {
-        when(item?.itemId){
+        when (item?.itemId) {
```
Comment (onstonboy) [inline app/src/main/java/com/ccc/chatapp/screens/chat/ChatActivity.kt]: 1 case thì dùng if luôn đi

File: app/src/main/java/com/ccc/chatapp/screens/login/LoginActivity.kt [hunk 4]
```diff
@@ -34,6 +34,10 @@ class LoginActivity : AppCompatActivity(), LoginView {
         handleEvents()
     }
 
+    override fun onSIgnIn() {
```
Comment (onstonboy) [inline app/src/main/java/com/ccc/chatapp/screens/login/LoginActivity.kt]: onSIgnIn -> onSignIn

File: app/src/main/java/com/ccc/chatapp/screens/signin/SignInActivity.kt [hunk 5]
```diff
@@ -1,13 +1,107 @@
...
+    }
+
+
+    private fun handleEvent() {
+        signInAccountButton.setOnClickListener { onSignInClick() }
+        avatarSignInImageView.setOnClickListener { pickImage() }
+    }
+
+    private fun onSignInClick() {
+        when (null) {
```
Comment (onstonboy) [inline app/src/main/java/com/ccc/chatapp/screens/signin/SignInActivity.kt]: (y) sáng tạo đó

File: app/src/main/java/com/ccc/chatapp/AppComponent.kt [hunk 6]
```diff
@@ -13,4 +14,6 @@ interface AppComponent {
     fun inject(loginActivity: LoginActivity)
 
     fun inject(chatActivity: ChatActivity)
+
+    fun inject(sinInActivity: SignInActivity)
```
Comment (onstonboy) [inline app/src/main/java/com/ccc/chatapp/AppComponent.kt]: sinInActivity -> signInActivity

File: app/src/main/java/com/ccc/chatapp/screens/login/LoginPresenter.kt [hunk 7]
```diff
@@ -28,6 +30,7 @@ class LoginPresenterImpl(
     }
 
     override fun signIn() {
+        view?.onSIgnIn()
```
Comment (onstonboy) [inline app/src/main/java/com/ccc/chatapp/screens/login/LoginPresenter.kt]: onSIgnIn -> onSignIn

File: app/src/main/java/com/ccc/chatapp/screens/signIn/SignInContract.kt [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package com.ccc.chatapp.screens.signin
+
+import com.ccc.chatapp.data.model.User
+import com.ccc.chatapp.screens.BasePresenter
+
+interface SigInView {
+    fun onSignInSuccess()
+    fun onSIgnInFailed()
```
Comment (onstonboy) [inline app/src/main/java/com/ccc/chatapp/screens/signIn/SignInContract.kt]: onSIgnInFailed ->onSignInFailed

File: app/src/main/java/com/ccc/chatapp/screens/signIn/SignInContract.kt [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+}
+
+interface SigInPresenter : BasePresenter {
```
Comment (onstonboy) [inline app/src/main/java/com/ccc/chatapp/screens/signIn/SignInContract.kt]: SigInPresenter -> SignInPresenter

File: app/src/main/java/com/ccc/chatapp/screens/signIn/SignInPresenter.kt [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+            .subscribeOn(schedulerProvider.io())
+            .observeOn(schedulerProvider.ui())
+            .subscribe({
+                view?.onSignInSuccess()
+            }, {
+                view?.onSIgnInFailed()
+                Log.e(Constant.Tag.SignInPresenter, it.message)
+            })
+        mCompositeDisposable.add(disposable)
+    }
```
Comment (onstonboy) [inline app/src/main/java/com/ccc/chatapp/screens/signIn/SignInPresenter.kt]: move this func to line 41

File: app/src/main/java/com/ccc/chatapp/screens/signin/SignInActivity.kt [hunk 11]
```diff
@@ -1,13 +1,106 @@
...
+    lateinit var mSchedulerProvider: SchedulerProvider
 
-class SignInActivity : AppCompatActivity() {
+    private lateinit var mPresenter: SigInPresenter
+    private lateinit var mSignInNavigator: SignInNavigator
+    private var mUser = User(
+        "",
+        "",
+        "",
+        Uri.parse("android.resource://com.ccc.chatapp/" + R.drawable.default_avatar).toString()
```
Comment (onstonboy) [inline app/src/main/java/com/ccc/chatapp/screens/signin/SignInActivity.kt]: "android.resource://com.ccc.chatapp/" + R.drawable.default_avatar -> hard code. create an Utils contain it.
Example: 
object Constant {
    const val BASE_URL = "android.resource://com.ccc.chatapp/"
}

File: app/src/main/java/com/ccc/chatapp/utils/Constant.kt [hunk 12]
```diff
@@ -2,14 +2,25 @@ package com.ccc.chatapp.utils
 
 object Constant {
 
+    object Tag {
+        const val LoginPresenter = "loginPresenter"
```
Comment (truyenz) [inline app/src/main/java/com/ccc/chatapp/utils/Constant.kt]: mấy cái TAG này dùng riêng cho từng activity thì nên viết trong activity luôn e

File: app/src/main/java/com/ccc/chatapp/screens/chat/ChatNavigator.kt [hunk 13]
```diff
@@ -1,12 +1,22 @@
...
+    fun goToLoginScreen()
     fun finishActivity()
 }
 
 class ChatNavigatorImpl(activity: ChatActivity) : BaseNavigator(activity), ChatNavigator {
+    override fun goToLoginScreen() {
+        val intent = Intent(activity, LoginActivity::class.java)
+        intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
+        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+        activity.startActivity(intent)
```
Comment (truyenz) [inline app/src/main/java/com/ccc/chatapp/screens/chat/ChatNavigator.kt]: write function getInstance() in LoginActivity

**Pull request merged**